packages/filebrowser/src/listing.ts

```diff
@@ -1123,24 +1123,59 @@ export class DirListing extends Widget {
     if (!files || files.length === 0) {
       return;
     }
-    const length = event.dataTransfer?.items.length;
-    if (!length) {
-      return;
-    }
-    for (let i = 0; i < length; i++) {
-      let entry = event.dataTransfer?.items[i].webkitGetAsEntry();
-      if (entry.isDirectory) {
-        console.log('currently not supporting drag + drop for folders');
-        void showDialog({
-          title: this._trans.__('Error Uploading Folder'),
-          body: this._trans.__(
-            'Drag and Drop is currently not supported for folders'
-          ),
-          buttons: [Dialog.cancelButton({ label: this._trans.__('Close') })]
-        });
+    event.preventDefault();
+
+    if (event.dataTransfer && event.dataTransfer.items) {
+      let items = event.dataTransfer.items;
+
+      const addDirectory = (item: any) => {
+        if (item.isDirectory) {
+          this._model.manager
+            .newUntitled({
+              path: this._model.path,
+              type: 'directory'
+            })
+            .then(async model => {
+              await this._manager.rename(
+                `${this._model.path}/${model.name}`,
+                `${this._model.path}/${item.name}`
+              );
+              this._model
+                .cd(`${this._model.path}/${item.name}`)
+                .catch(error =>
+                  showErrorMessage(
+                    this._trans._p('showErrorMessage', 'Open directory'),
+                    error
+                  )
+                );
+            })
+            .catch(err => {
+              console.log('error while creating folder: ', err);
+            });
+
+          let directoryReader = item.createReader();
+
+          directoryReader.readEntries((entries: any) => {
+            entries.forEach((entry: any) => {
+              addDirectory(entry);
+            });
+          });
```

Comment:
I think we should be only adding files that are inside the directory after we created, renamed and entered it so in `then()` after `.cd()`.

Reply:
I  tried the adding `.then()` to `cd()` function. It did rename the directory name as expected, but didn't upload files in the correct sub-directory,

Reply:
I guess I would add a second argument `path` to the `addDirectory` function and try to ensure that the path is set and updated correctly at each step. I highly recommend walking the execution step-by-step with a debugger (e.g. in Chrome developer tools).

Reply:
Okay, I'll try that. 👍🏻

Reply:
I tried this approach. And it is fixing the issue up to a great extent. But it is encountering a problem with the nested file in a nested folder, otherwise, it can upload nested files up to 1 level.